### PR TITLE
[WIP] Unified csp containers

### PR DESCRIPTION
**Pull request by drobison00**
Work in progress, initial work toward HPO container unification.

Note: It looks like more files have been changed than really have.

aws/azure/gcp folders moved under a top level cloud_service_providers folder, to give access to the top level docker build file.

### CODE REVIEW

**Comment by miroenev** [comment]
@drobison00 I think this is a good start!
In the case of the AWS HPO demo, we do a container build from the notebook instance. Would it be possible to disable non AWS build steps, or would you suggest maintaining a minimal cloud specific version for this purpose?

**Comment by drobison00** [comment]
@miroenev I don't think we can disable sections directly, although I can think of some possible ways to do this with multi-stage builds, or using underlying configuration scripts that can handle flow control; in either case I think if we start down that road there isn't much value left from taking a unified container approach. With the unified approach we're going to sacrifice in terms of container size to get general applicability.

**Comment by zronaghi** [comment]
For Azure, we can modify the notebook to use a Docker image by replacing `environment_definition` with `custom_docker_image` in the Estimator. AzureML will also create a snapshot of the local `source_directory`, where we copy the training scripts. Just tested the Dockerfile and works well without the entrypoint script (ENTRYPOINT [ "/bin/bash" ]), attached the logs

[azureml-csp-container-working.txt](https://github.com/rapidsai/cloud-ml-examples/files/5570358/azureml-csp-container-working.txt)
[azureml-csp-container-failing.txt](https://github.com/rapidsai/cloud-ml-examples/files/5570359/azureml-csp-container-failing.txt)

File: cloud_service_providers/Dockerfile.training.unified [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        sqlalchemy \
+        'ray[tune]'
+
+RUN mkdir -p /opt/rapids_cloudml
+ADD infrastructure/* \
+     /opt/rapids_cloudml/
+
+## Unified entrypoint
+WORKDIR "/opt/rapids_cloudml"
+ENTRYPOINT [ "bash", "entrypoint.sh" ]
```
Comment (zronaghi) [inline cloud_service_providers/Dockerfile.training.unified:91]:
```suggestion
RUN chmod +x /opt/rapids_cloudml/entrypoint.sh
ENTRYPOINT [ "/usr/bin/tini", "--", "/opt/rapids_cloudml/entrypoint.sh" ]

CMD [ "/bin/bash" ]
```

File: cloud_service_providers/infrastructure/entrypoint.sh [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  echo "Running: entrypoint.py ${ARGS[@]}"
+  python entrypoint.py ${ARGS[@]}
+
+else
+  # Azure
+  # TODO: Azure workflow is substantially different.
+  echo "Running AzureML HPO entrypoint."
+  cd /opt/rapids/azure || exit 1
+  echo "Running: ${ARGS[@]}"
+  eval "${ARGS[@]}"
```
Comment (zronaghi) [inline cloud_service_providers/infrastructure/entrypoint.sh:44]:
```suggestion
 exec "$@"
```

File: cloud_service_providers/infrastructure/entrypoint.sh [hunk 3]
```diff
@@ -0,0 +1,45 @@
+source /conda/etc/profile.d/conda.sh
```
Comment (zronaghi) [inline cloud_service_providers/infrastructure/entrypoint.sh:1]:
```suggestion
#!/bin/bash

source /conda/etc/profile.d/conda.sh
```

File: cloud_service_providers/Dockerfile.training.unified (hunk 1 above)
Comment (zronaghi) [inline cloud_service_providers/Dockerfile.training.unified:91]:
```suggestion
RUN chmod +x /opt/rapids_cloudml/entrypoint.sh
ENTRYPOINT [ "/usr/bin/tini", "--", "/opt/rapids_cloudml/entrypoint.sh" ]

CMD [ "/bin/bash" ]
```

File: cloud_service_providers/infrastructure/entrypoint.sh (hunk 3 above)
Comment (zronaghi) [inline cloud_service_providers/infrastructure/entrypoint.sh:1]:
```suggestion
#!/bin/bash

source /conda/etc/profile.d/conda.sh
```

File: cloud_service_providers/infrastructure/entrypoint.sh (hunk 2 above)
Comment (zronaghi) [inline cloud_service_providers/infrastructure/entrypoint.sh:44]:
```suggestion
  exec "$@"
```

**Review by zronaghi — COMMENTED** [review]
Added few suggestions to modify the entryscript, that works on Azure

File: common/docker/Dockerfile.training.unified [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+ADD azure/* \
+    /opt/rapids/azure/
+
+RUN . /opt/conda/etc/profile.d/conda.sh \
+    && conda activate rapids \
+    && pip install \
+        azureml-sdk \
+        azureml-widgets \
+        azureml-mlflow
+
```
Comment (Nanthini10) [inline common/docker/Dockerfile.training.unified:70]: Should we also include optuna and (dask_optuna) installation here for the future demos?

Comment (zronaghi) [reply common/docker/Dockerfile.training.unified:70]: Yes, let's add these for the azure-optuna demo

**Comment by zronaghi** [comment]
@drobison00 the modified Dockerfile and entrypoint still fails on Azure, only the suggested version above runs successfully

**Comment by drobison00** [comment]
@zronaghi Looking at your azure logs, they're passing some very complicated commands that relay on a handful of mounted volumes, this includes a great deal of quoted text and 'bash -c' commands. I'll need to experiment a bit and see what could be breaking the raw command pass-through.

File: common/docker/Dockerfile.training.unified [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        sqlalchemy \
+        'ray[tune]'
+
+RUN mkdir -p /opt/rapids_cloudml
+ADD common/docker/infrastructure/* \
+     /opt/rapids_cloudml/
+
+## Unified entrypoint
+WORKDIR "/opt/rapids_cloudml"
+ENTRYPOINT [ "bash", "entrypoint.sh" ]
```
Comment (zronaghi) [inline common/docker/Dockerfile.training.unified:93]: Adding /path/to/entrypoint fixes the issue:

```suggestion
ENTRYPOINT [ "bash", "/opt/rapids_cloudml/entrypoint.sh" ]
```
Looking at the docker run command from logs, -w overrides the Dockerfile WORKDIR

**Review by zronaghi — APPROVED** [review]
Looks great!

**Pull request merged**